File: emulinker/src/main/java/org/emulinker/kaillera/model/KailleraUser.java
package org.emulinker.kaillera.model;

import java.net.InetSocketAddress;
import java.util.Collection;
import org.emulinker.kaillera.model.event.KailleraEventListener;
import org.emulinker.kaillera.model.exception.*;
import org.emulinker.kaillera.model.impl.KailleraGameImpl;
import org.emulinker.kaillera.model.impl.KailleraUserImpl;

public interface KailleraUser {
  public static final byte CONNECTION_TYPE_LAN = 1;
  public static final byte CONNECTION_TYPE_EXCELLENT = 2;
  public static final byte CONNECTION_TYPE_GOOD = 3;
  public static final byte CONNECTION_TYPE_AVERAGE = 4;
  public static final byte CONNECTION_TYPE_LOW = 5;
  public static final byte CONNECTION_TYPE_BAD = 6;

  public static final String[] CONNECTION_TYPE_NAMES = {
    "DISABLED", "LAN", "Excellent", "Good", "Average", "Low", "Bad"
  };

  public static final byte STATUS_PLAYING = 0;
  public static final byte STATUS_IDLE = 1;
  public static final byte STATUS_CONNECTING = 2;
  public static final String[] STATUS_NAMES = {"Playing", "Idle", "Connecting"};

  public int getID();

  public InetSocketAddress getConnectSocketAddress();

  public String getProtocol();

  public long getConnectTime();

  public int getFrameCount();

  public byte[] getLostInput();

  public int getArraySize();

  public int getBytesPerAction();

  public void setTempDelay(int tempDelay);

  public void setFrameCount(int frameCount);

  public boolean getMsg();

  public void setMsg(boolean msg);

  public int getStatus();

  public void setLastMsgID(int lastMsgID);

  public int getDelay();

  public int getLastMsgID();

  public boolean getP2P();

  public int getTimeouts();

  public void setTimeouts(int timeouts);

  public boolean findIgnoredUser(String address);

  public void setP2P(boolean p2P);

  public void setPlayerNumber(int playerNumber);

  public boolean getMute();

  public void setMute(boolean mute);

  public KailleraGameImpl getGame();

  public int getAccess();

  public boolean removeIgnoredUser(String address, boolean removeAll);

  public boolean searchIgnoredUsers(String address);

  public void setIgnoreAll(boolean ignoreAll);

  public boolean getIgnoreAll();

  public void addIgnoredUser(String address);

  public Collection<KailleraUserImpl> getUsers();

  public String getName();

  public void setName(String name);

  public boolean getStealth();

  public void setStealth(boolean stealth);

  public String getClientType();

  public boolean isEmuLinkerClient();

  public void setClientType(String clientType);

  public byte getConnectionType();

  public void setConnectionType(byte connectionType);

  public InetSocketAddress getSocketAddress();

  public void setSocketAddress(InetSocketAddress clientSocketAddress);

  public int getPing();

  public void setPing(int ping);

  public void login()
      throws PingTimeException, ClientAddressException, ConnectionTypeException, UserNameException,
          LoginException;

  public long getLastActivity();

  public void updateLastActivity();

  public void updateLastKeepAlive();

  public long getLastKeepAlive();

  public boolean isLoggedIn();

  public KailleraServer getServer();

  public KailleraEventListener getListener();

  public void chat(String message) throws ChatException, FloodException;

  public KailleraGame createGame(String romName) throws CreateGameException, FloodException;

  public void quit(String message)
      throws QuitException, DropGameException, QuitGameException, CloseGameException;

  public KailleraGame joinGame(int gameID) throws JoinGameException;

  public int getPlayerNumber();

  public void startGame() throws StartGameException;

  public void gameChat(String message, int messageID) throws GameChatException;

  public void gameKick(int userID) throws GameKickException;

  public void playerReady() throws UserReadyException;

  public void addGameData(byte[] data) throws GameDataException;

  public void dropGame() throws DropGameException;

  public void quitGame() throws DropGameException, QuitGameException, CloseGameException;

  public void droppedPacket();

  public void stop();
}

File: emulinker/src/main/java/org/emulinker/kaillera/model/KailleraUser.kt
package org.emulinker.kaillera.model

import java.net.InetSocketAddress
import java.time.Instant
import kotlin.Throws
import kotlinx.coroutines.sync.Mutex
import org.emulinker.kaillera.model.event.KailleraEventListener
import org.emulinker.kaillera.model.exception.*
import org.emulinker.kaillera.model.impl.KailleraGameImpl
import org.emulinker.kaillera.model.impl.KailleraUserImpl

interface KailleraUser {
  // Fields that only support getters.
  val id: Int
  /**
   * Level of access that the user has.
   *
   * See AdminCommandAction for available values. This should be turned into an enum.
   */
  val accessLevel: Int
  /** User action data response message size (in number of bytes). */
  val arraySize: Int
  val bytesPerAction: Int
  val connectSocketAddress: InetSocketAddress
  val connectTime: Instant
  val frameDelay: Int
  val game: KailleraGameImpl?
  val isEmuLinkerClient: Boolean
  val lastActivity: Instant
  val lastKeepAlive: Instant
  val listener: KailleraEventListener
  val loggedIn: Boolean
  val protocol: String
  val server: KailleraServer
  val status: UserStatus
  val users: Collection<KailleraUserImpl?>?

  // Fields with public getters and setters.
  var clientType: String?
  var connectionType: ConnectionType
  var frameCount: Int
  var ignoreAll: Boolean
  var lastMsgID: Int
  var isAcceptingDirectMessages: Boolean
  var isMuted: Boolean
  var name: String? // TODO(nue): Remove this "?"
  /**
   * This is called "p2p mode" in the code and commands.
   *
   * See the command /p2pon.
   */
  var ignoringUnnecessaryServerActivity: Boolean
  var ping: Int
  var playerNumber: Int // TODO(nue): Make this nullable.
  var socketAddress: InetSocketAddress
  var inStealthMode: Boolean
  var tempDelay: Int
  var timeouts: Int
  var smallLagSpikesCausedByUser: Long
  var bigLagSpikesCausedByUser: Long

  // Methods.
  @Throws(
      PingTimeException::class,
      ClientAddressException::class,
      ConnectionTypeException::class,
      UserNameException::class,
      LoginException::class)
  suspend fun login()

  fun updateLastActivity()

  fun updateLastKeepAlive()

  fun getLostInput(): ByteArray

  fun findIgnoredUser(address: String): Boolean

  fun removeIgnoredUser(address: String, removeAll: Boolean): Boolean

  fun searchIgnoredUsers(address: String): Boolean

  fun addIgnoredUser(address: String)

  @Throws(ChatException::class, FloodException::class)
  fun chat(message: String)

  @Throws(CreateGameException::class, FloodException::class)
  suspend fun createGame(romName: String): KailleraGame

  @Throws(
      QuitException::class,
      DropGameException::class,
      QuitGameException::class,
      CloseGameException::class)
  fun quit(message: String?)

  @Throws(JoinGameException::class)
  suspend fun joinGame(gameID: Int): KailleraGame

  @Throws(StartGameException::class)
  fun startGame()

  @Throws(GameChatException::class)
  fun gameChat(message: String, messageID: Int)

  @Throws(GameKickException::class)
  fun gameKick(userID: Int)

  @Throws(UserReadyException::class)
  fun playerReady()

  @Throws(GameDataException::class)
  fun addGameData(data: ByteArray)

  @Throws(DropGameException::class)
  fun dropGame()

  @Throws(DropGameException::class, QuitGameException::class, CloseGameException::class)
  fun quitGame()

  fun droppedPacket()

  suspend fun stop()
  val mutex: Mutex
}
